### DISCUSSION TITLE
Single publisher with high publish rate -> message_queue_too_long

### DISCUSSION THREAD

**Discussion by freol35241** [emqx/emqx#11212] [Q&A]
Hi,

Trial setup:
EMQX 5.1.0
Single publisher connected via websocket listener publishing **QoS0** messages in a loop
3 Shared subscribers (subscribing using **QoS0**)

I have started to run into some issues where the publishing client get forcefully disconnected (i.e. the websocket connection is terminated) with the following error [debug?] message:

```
[debug] msg: websocket_terminated, mfa: emqx_ws_connection:terminate/3, line: 489, peername: *****, clientid: *********, reason: {shutdown,#{max => 1000,reason => message_queue_too_long,value => 1911}}, tag: SOCKET
```

I think this log message actually relates to the erlang mailbox of the receiving process is full rather than the `mqueue` used for QoS1 and QoS2 messages. Is that correct?

Yes, I am aware of the `force_shutdown` configuration options (https://www.emqx.io/docs/en/v5.1/configuration/configuration-manual.html#force-shutdown). However, disabling this seems to allow the mailbox to grow indefinitely which is really not a good option in my case. At one point the OS will simply kill the process due to memory shortage?

I have tried to make use of the rate limiting feature of emqx to put some back pressure on the publisher but it seems not to work as I expect as I get the same error message (above) and forcefully disconnected clients in the same manner. It seems that the back pressure is only applied between the "main" emqx process and the "receiving" erlang process and does not propagate to the websocket connection to the publisher?

I would really appreciate some help in understanding:
* What kind of performance can i expect from EMQX in a scenario like this? What is the expected throughput in messages/s for small messages (<100bytes each). And where is the bottleneck?
* Can I make use of the rate limiting feature to put backpressure on the publisher, thus avoiding filling up the mailbox? How?
* How should I think about the `force_shutdown` options?

**Comment by qzhuyan** [comment]
> I think this log message actually relates to the erlang mailbox of the receiving process is full rather than the mqueue used for QoS1 and QoS2 messages. Is that correct?
Yes. Both publisher connecion process or subscriber connection process may get long msg queue. the msgq could contain transport packets, signals from other processes and MQTT messages.  Could you check if it is subscriber or publisher process get killed?

You may also check what is in the process msgq when it starts grow.

> What kind of performance can i expect from EMQX in a scenario like this? What is the expected throughput in messages/s for small messages (<100bytes each). And where is the bottleneck?

We cannot tell, every cluster has different types of workload with different enabled features. Only test in staging env could tell.

> Can I make use of the rate limiting feature to put backpressure on the publisher, thus avoiding filling up the mailbox? How?

yes, see https://www.emqx.io/docs/en/v5.1/rate-limit/rate-limit.html#limiter-types

limiter configs are under `listeners`

https://www.emqx.io/docs/en/v5.1/configuration/limiter.html

> How should I think about the force_shutdown options?

The force shutdown is designed to avoid single client process consume too many memory and CPU that effect the overall system performance since heavy memory allocation is not preemptable . If your clients are well managed you could turn it off.

**Reply by freol35241** [reply]
Thanks for your response @qzhuyan !

It is the publisher process that gets killed. How can I check what's in the process msgqueue when it gets killed? 

Currently, the setup is a single instance of emqx. The number of clients are not expected to be more than ~50, would I still benefit from a multi-node setup you think?

I have tried configuring the rate limiter accordingly but I keep running in to the same problem/error (I.e. the publisher gets kicked out due to too long msg queue). When introducing the rate limiter, it seems like the publishing client is still allowed publish messages at whatever rate but the subscribers "see" the expected, limited rate according to the limiter config. Any thoughts on this?

**Reply by qzhuyan** [reply]
you could monitor it with
```
emqx  eval  'observer_cli:start()'
```

then enter 'mq' to sort the process with length of msg queue.

**Reply by qzhuyan** [reply]
will be helpful If we could get a snapshot of the `observer_cli` output.

once you find the trouble process you could dump the messages in the msgq.

``` bash
emqx remote_console
```

Then in emqx console
```
process_info(Pid, messages).
```

**Reply by freol35241** [reply]
Sorry for the long delay. I had a go at this with the following results:
![image](https://github.com/emqx/emqx/assets/21332885/c650840d-b940-47df-9d55-b182a34c19a6)

Unfortunately, the message queue grows so quickly (and the process gets killed so quickly) that I haven't managed to dump the messages... But I dont think the disconnections are related to the messages in the message queue, rather just the fact that the queue gets filled and the `force_shutdown` options therefore kicks that client out.

But, from the snapshot above, I realized that having debug logging turned on does not help the performance obviously. Setting the logging level to `warning`, I manage to set the rate limiter to `1000/s` and keep pushing messages without filling the message queue (`5000/s` filled the message queue with client disconnection as a result). 

However, when using a rate limit, I also noticed that the publisher process (i.e the client process) is done (i.e. has published all messages and exits) long before (minutes!) the broker processes them and forwards them to any subscribers. The subscribers receives these messages at the rate which was configured for the rate limiter so that is as expected. It seems the publisher pushes messages into some large buffer somewhere(?) and then exits and the broker receiver process reads from this buffer at the prescribed limited rate. But I dont really understand where this buffer resides? My best guess is that it is all the accumulated socket buffers "along the connection" between the client machine and the broker machine? Do you have any input to give here?

In conclusion, I think what I am missing is:
* a setting that will automatically rate limit a publisher if it publishes messages at such a high rate that it fills the message queue. Right now, I can either choose to not rate limit at all (and clients will get disconnected if publishing with too high rate) **or** rate limit explicitly using the current available rate limiting settings. 
* an understanding of the buffering behavior described above.

**Reply by lafirest** [reply]
@freol35241 
I think the session which was killed maybe not is the publisher but the subscriber.
At ideal, the rate limit will back pressure on the publisher, but now the behaviour does not perform well, the system kernel and the Erlang socket port both may buffer some data, thus our strategy is to endeavour to satisfy the rate claims by an internal buffer until the OOM strategy kills the session when it exceeds.

**Reply by lafirest** [reply]
It is possible to set the rate limiter to drop incoming data when overloaded, but it is an undocumented feature.
Just for your situation, I think more data need to be input, e.g. the total incoming rate of the publisher, in a normal scenario, if all messages are QoS 0, the publisher rarely accumulates too many messages in its message queue.

**Reply by freol35241** [reply]
Thanks for the response @lafirest . 

I am actually quite sure it is the receiving process that fills its message queue (or "erlang mailbox"), even using only QoS0 messages. The subscribers all work fine and only the publisher gets kicked out with a log message stating that the message queue got too long and therefore the websocket connection gets terminated. What do you define as "normal scenario"?

If I understand you correctly, the current implementation of the rate limiter does not actually put any back pressure on the publisher?

**Reply by lafirest** [reply]
Sorry I used a broad sentence "normal scenario", it just wants to clarify most time the session is killed due to it being a subscriber and accumulating many messages.
There are no precise metrics to measure how much the pressure back to the publisher, but it does not say this strategy has no effect, at least in my testing, the pressure can be feedback to the publisher eventually.
Can you provide more information, including the EMQX configuration, the CPU and Memory and the maximum publish rate of the publisher

**Reply by thalesmg** [reply]
As you probably know, if you combine QoS 1 or 2 messages with the `max_inflight`, publishers should have back pressure when publishing too fast.  If that's an option for your use case, you could give it a try.
